src/app/application/services/data-source/index.ts

```diff
@@ -1,12 +1,35 @@
 import CallableEventSource, { isClosed } from './CallableEventSource'
 import type { Creator, Destroyer } from './DataSourcePool'
 
+type ExtractRouteParams<T extends string> =
+  string extends T
+    ? Record<string, string>
+    : T extends `${infer _Start}:${infer Param}/${infer Rest}`
+      ? {[k in Param | keyof ExtractRouteParams<Rest>]: string}
+      : T extends `${infer _Start}:${infer Param}`
+        ? {[k in Param]: string}
+        : {};
+
+type PaginationParams = {
+  size: number
+  page: number
+  search: string
+}
+
+type ExtractSources<T extends string, K> = {
+  [Route in T]: (params: ExtractRouteParams<Route> & K, source: { close: () => void }) => void
```

Comment:
**Comment**: It’s a little unfortunate that this implies that `params` will always have `PaginationParams`.

Reply:
Yeah for sure defo agree there, that being said, theoretically they always could potentially have these I suppose. I think this was half my thinking with not putting these things in the URI for the DataSource in the first place. Overall still thinking about this one, but I get what you are saying here.

(btw, I'll probably change the name of PaginationParams at some point)

Reply:
Wanted to check if you had any more feedback on this one, if not we can decide whether we want to roll this out or not.

Reply:
Indifferent about this for now. In general we of course shouldn’t say `params` can potentially have these fields because that might not always be correct, but it might be good enough for now.